# Challenge


# 1. Knowledge domain
# 2. Exploratory analysis
# 3. Machine learning model
## 3.1. First approach
## 3.2. Second approach
# 4. Result of the competition
# References

------------------

En este notebook:

## 3.2. Second approach
### Knn
### Decision Tree
### Gaussian NB
### Random Forest
# 4. Result of the competition
# References

--------------

## 3.2. Second approach

### Cargo Valores

In [41]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

from pandas import Series, DataFrame

data = pd.read_csv('data/train.csv', sep=",", encoding = "ISO-8859-1")
df = data.copy()

data_test_nolabel = pd.read_csv('data/test_nolabel.csv', sep=",", encoding = "ISO-8859-1")
df_test_nolabel = data_test_nolabel.copy()

In [42]:
data_test_nolabel = pd.read_csv('data/test_nolabel.csv', sep=",", encoding = "ISO-8859-1")
df_test_nolabel = data_test_nolabel.copy()

### Cambios en Features

### DisbursementDate

In [43]:
# drop ChgOffDate column
#df.drop(['ChgOffDate'], axis=1, inplace=True)
df.drop(['State'], axis=1, inplace=True)
df.drop(['BalanceGross'], axis=1, inplace=True)
df.drop(['id'], axis=1, inplace=True)

# drop ChgOffDate column
#df_test_nolabel.drop(['ChgOffDate'], axis=1, inplace=True)
df_test_nolabel.drop(['State'], axis=1, inplace=True)
df_test_nolabel.drop(['BalanceGross'], axis=1, inplace=True)
#df_test_nolabel.drop(['id'], axis=1, inplace=True)

In [44]:
#Change the format into datetime
df['DisbursementDate'] = pd.to_datetime(data['DisbursementDate'])

#Change the format into datetime
df_test_nolabel['DisbursementDate'] = pd.to_datetime(data['DisbursementDate'])

In [45]:
df['DisbursementDay'] = df['DisbursementDate'].apply(lambda x: x.day)

df_test_nolabel['DisbursementDay'] = df_test_nolabel['DisbursementDate'].apply(lambda x: x.day)

In [46]:
df['DisbursementMonth'] = df['DisbursementDate'].apply(lambda x: x.month)
df_test_nolabel['DisbursementMonth'] = df_test_nolabel['DisbursementDate'].apply(lambda x: x.month)

In [47]:
df['DisbursementYear'] = df['DisbursementDate'].apply(lambda x: x.year)
df_test_nolabel['DisbursementYear'] = df_test_nolabel['DisbursementDate'].apply(lambda x: x.year)

In [48]:
# We can now delete the 'DisbursementDate' column.
df.drop(['DisbursementDate'], axis=1, inplace=True)
# We can now delete the 'DisbursementDate' column.
df_test_nolabel.drop(['DisbursementDate'], axis=1, inplace=True)

### ApprovalDate

In [49]:
#Change the format into datetime
df['ApprovalDate'] = pd.to_datetime(data['ApprovalDate'])
#Change the format into datetime
df_test_nolabel['ApprovalDate'] = pd.to_datetime(data['ApprovalDate'])

In [50]:
df['ApprovalDay'] = df['ApprovalDate'].apply(lambda x: x.day)
df['ApprovalMonth'] = df['ApprovalDate'].apply(lambda x: x.month)
df['ApprovalYear'] = df['ApprovalDate'].apply(lambda x: x.year)

df_test_nolabel['ApprovalDay'] = df_test_nolabel['ApprovalDate'].apply(lambda x: x.day)
df_test_nolabel['ApprovalMonth'] = df_test_nolabel['ApprovalDate'].apply(lambda x: x.month)
df_test_nolabel['ApprovalYear'] = df_test_nolabel['ApprovalDate'].apply(lambda x: x.year)

In [51]:
# We can now delete the 'ApprovalDate' column.
df.drop(['ApprovalDate'], axis=1, inplace=True)
# We can now delete the 'ApprovalDate' column.
df_test_nolabel.drop(['ApprovalDate'], axis=1, inplace=True)

### DisbursementGross

In [52]:
def clean_currency(x):
   
    if isinstance(x, str):
        return(x.replace('$', '').replace(',', ''))
    return(x)

In [53]:
df['DisbursementGross'] = df['DisbursementGross'].apply(clean_currency).astype('float')
df_test_nolabel['DisbursementGross'] = df_test_nolabel['DisbursementGross'].apply(clean_currency).astype('float')

### ChgOffDate

In [54]:
df['ChgOffDate'] = df['ChgOffDate'].fillna(1)

for i in range(len(df)):
    if df['ChgOffDate'][i] != 1 :
        df['ChgOffDate'][i] = 0   

df['ChgOffDate'] = df['ChgOffDate'].astype(int)

df_test_nolabel['ChgOffDate'] = df_test_nolabel['ChgOffDate'].fillna(1)

for i in range(len(df_test_nolabel)):
    if df_test_nolabel['ChgOffDate'][i] != 1 :
        df_test_nolabel['ChgOffDate'][i] = 0   
        
df_test_nolabel['ChgOffDate'] = df_test_nolabel['ChgOffDate'].astype(int)

In [55]:
df.head(10)

## Name

In [ ]:
name = []

lista_1 = dict()
for i in range (0, 49320): 
    name.append(df['Name'][i])
    
for j in range (0, 49320):
        
    if df['Name'][j] not in lista_1: 
        c = name[j]
        df['Name'][j] = j
        lista_1[c] = j
        print(c)
        print(df['Name'][j])
        print(lista_1)
        
    else :
        df['Name'][j] = lista_1[df['Name'][j]]

In [ ]:
name = []

lista_1 = dict()
for i in range (0, 7994): 
    name.append(df_test_nolabel['Name'][i])
    
for j in range (0, 7994):
        
    if df_test_nolabel['Name'][j] not in lista_1: 
        c = name[j]
        df_test_nolabel['Name'][j] = j
        lista_1[c] = j
        print(c)
        print(df_test_nolabel['Name'][j])
        print(lista_1)
        
    else :
        df_test_nolabel['Name'][j] = lista_1[df_test_nolabel['Name'][j]]

## Bank

In [ ]:
name_bank = []
lista = dict()

for i in range (0, 49320):
    name_bank.append(df['Bank'][i])
    
for j in range (0, 49320):
    
    if df['Bank'][j] not in lista:
        c = name_bank[j]
        df['Bank'][j] = j
        lista[c] = j
        print(c)
        print(df['Bank'][j])
        print(lista)
        
    else :
        df['Bank'][j] = lista[df['Bank'][j]]
        
name_bank = []
lista = dict()

for i in range (0, 7994):
    name_bank.append(df_test_nolabel['Bank'][i])
    
for j in range (0, 7994):
    
    if df_test_nolabel['Bank'][j] not in lista:
        c = name_bank[j]
        df_test_nolabel['Bank'][j] = j
        lista[c] = j
        print(c)
        print(df_test_nolabel['Bank'][j])
        print(lista)
        
    else :
        df_test_nolabel['Bank'][j] = lista[df_test_nolabel['Bank'][j]]

In [ ]:
City = []
lista_2 = dict()

for i in range (0, 49320):
    City.append(df['City'][i])
    
for j in range (0, 49320):
    
    if df['City'][j] not in lista_2:
        c = City[j]
        df['City'][j] = j
        lista_2[c] = j
        print(c)
        print(df['City'][j])
        print(lista_2)
        
    else :
        df['City'][j] = lista_2[df['City'][j]]
        

City = []
lista_2 = dict()

for i in range (0, 7994):
    City.append(df_test_nolabel['City'][i])
    
for j in range (0, 7994):
    
    if df_test_nolabel['City'][j] not in lista_2:
        c = City[j]
        df_test_nolabel['City'][j] = j
        lista_2[c] = j
        print(c)
        print(df_test_nolabel['City'][j])
        print(lista_2)
        
    else :
        df_test_nolabel['City'][j] = lista_2[df_test_nolabel['City'][j]]

## BankState

In [ ]:
BankState = []
lista_4 = dict()

for i in range (0, 49320):
    BankState.append(df['BankState'][i])
    
for j in range (0, 49320):

    if df['BankState'][j] not in lista_4:
        c = BankState[j]
        df['BankState'][j] = j
        lista_4[c] = j
        print(c)
        print(df['BankState'][j])
        print(lista_4)
        
    else :
        df['BankState'][j] = lista_4[df['BankState'][j]]
        
BankState = []
lista_4 = dict()

for i in range (0, 7994):
    BankState.append(df_test_nolabel['BankState'][i])
    
for j in range (0, 7994):

    if df_test_nolabel['BankState'][j] not in lista_4:
        c = BankState[j]
        df_test_nolabel['BankState'][j] = j
        lista_4[c] = j
        print(c)
        print(df_test_nolabel['BankState'][j])
        print(lista_4)
        
    else :
        df_test_nolabel['BankState'][j] = lista_4[df_test_nolabel['BankState'][j]]

## ApprovalFY

In [ ]:
ApprovalFY = []
lista_6 = dict()

for i in range (0, 49320):
    ApprovalFY.append(df['ApprovalFY'][i])
    
for j in range (0, 49320):
    
    if df['ApprovalFY'][j] not in lista_6:
        c = ApprovalFY[j]
        df['ApprovalFY'][j] = j
        lista_6[c] = j
        print(c)
        print(df['ApprovalFY'][j])
        print(lista_6)
        
    else :
        df['ApprovalFY'][j] = lista_6[df['ApprovalFY'][j]]
        
ApprovalFY = []
lista_6 = dict()

for i in range (0, 7994):
    ApprovalFY.append(df_test_nolabel['ApprovalFY'][i])
    
for j in range (0, 7994):
    
    if df_test_nolabel['ApprovalFY'][j] not in lista_6:
        c = ApprovalFY[j]
        df_test_nolabel['ApprovalFY'][j] = j
        lista_6[c] = j
        print(c)
        print(df_test_nolabel['ApprovalFY'][j])
        print(lista_6)
        
    else :
        df_test_nolabel['ApprovalFY'][j] = lista_6[df_test_nolabel['ApprovalFY'][j]]

## RevLineCr

In [ ]:
RevLineCr = []
lista_7 = dict()

for i in range (0, 49320):
    RevLineCr.append(df['RevLineCr'][i])
    
for j in range (0, 49320):
        
    if df['RevLineCr'][j] not in lista_7:
        c = RevLineCr[j]
        df['RevLineCr'][j] = j
        lista_7[c] = j
        print(c)
        print(df['RevLineCr'][j])
        print(lista_7)
        
    else :
        df['RevLineCr'][j] = lista_7[df['RevLineCr'][j]]
        
RevLineCr = []
lista_7 = dict()

for i in range (0, 7994):
    RevLineCr.append(df_test_nolabel['RevLineCr'][i])
    
for j in range (0, 7994):
        
    if df_test_nolabel['RevLineCr'][j] not in lista_7:
        c = RevLineCr[j]
        df_test_nolabel['RevLineCr'][j] = j
        lista_7[c] = j
        print(c)
        print(df_test_nolabel['RevLineCr'][j])
        print(lista_7)
        
    else :
        df_test_nolabel['RevLineCr'][j] = lista_7[df_test_nolabel['RevLineCr'][j]]

## LowDoc

In [ ]:
LowDoc = []
lista_8 = dict()

for i in range (0, 49320):
    LowDoc.append(df['LowDoc'][i])
    
for j in range (0, 49320):
        
    if df['LowDoc'][j] not in lista_8:
        c = LowDoc[j]
        df['LowDoc'][j] = j
        lista_8[c] = j
        print(c)
        print(df['LowDoc'][j])
        print(lista_8)
        
    else :
        df['LowDoc'][j] = lista_8[df['LowDoc'][j]]
        
LowDoc = []
lista_8 = dict()

for i in range (0, 7994):
    LowDoc.append(df_test_nolabel['LowDoc'][i])
    
for j in range (0, 7994):
        
    if df_test_nolabel['LowDoc'][j] not in lista_8:
        c = LowDoc[j]
        df_test_nolabel['LowDoc'][j] = j
        lista_8[c] = j
        print(c)
        print(df_test_nolabel['LowDoc'][j])
        print(lista_8)
        
    else :
        df_test_nolabel['LowDoc'][j] = lista_8[df_test_nolabel['LowDoc'][j]]

In [ ]:
df['Name'] = df['Name'].astype(float)
df['City'] = df['City'].astype(float)
df['Bank'] = df['Bank'].astype(float)
df['BankState'] = df['BankState'].astype(float)
df['ApprovalFY'] = df['ApprovalFY'].astype(float)
df['RevLineCr'] = df['RevLineCr'].astype(float)
df['LowDoc'] = df['LowDoc'].astype(float)

df_test_nolabel['Name'] = df_test_nolabel['Name'].astype(float)
df_test_nolabel['City'] = df_test_nolabel['City'].astype(float)
df_test_nolabel['Bank'] = df_test_nolabel['Bank'].astype(float)
df_test_nolabel['BankState'] = df_test_nolabel['BankState'].astype(float)
df_test_nolabel['ApprovalFY'] = df_test_nolabel['ApprovalFY'].astype(float)
df_test_nolabel['RevLineCr'] = df_test_nolabel['RevLineCr'].astype(float)
df_test_nolabel['LowDoc'] = df_test_nolabel['LowDoc'].astype(float)

In [ ]:
df.dtypes

In [ ]:
df_test_nolabel.dtypes

In [ ]:
(df_test_nolabel.isna().sum()/df_test_nolabel.shape[0]).sort_values(ascending=True)

## Dropeo de NaN

In [ ]:
# Delete the Nan of the variables which were of type float/int since the beginning
# NO HACER EN EL TEST
df.dropna(axis=0, inplace=True)

In [ ]:
df_test_nolabel.shape

In [ ]:
indexNames = df[ df['LowDoc'] == 32 ].index
# Delete these row indexes from dataFrame
df.drop(indexNames , inplace=True)

In [ ]:
indexNames = df[ df['BankState'] == 946 ].index
# Delete these row indexes from dataFrame
df.drop(indexNames , inplace=True)

In [ ]:
indexNames = df[ df['RevLineCr'] == 145].index
# Delete these row indexes from dataFrame
df.drop(indexNames , inplace=True)

In [ ]:
train = df

def corr_heatmap(train):
    correlations = train.corr()

    fig, ax = plt.subplots(figsize=(20,20))
    sns.heatmap(correlations, vmax=0.1, center=0, fmt='.2f',cmap="PiYG",
                square=True, linewidths=.5, annot=True, cbar_kws={"shrink": 0.9})
    plt.show();
    
corr_heatmap(train)

## Machine Learning

### KNN: Training

In [ ]:
#Train Test Split
#Use train_test_split to split your data into a training set and a testing set
from sklearn.model_selection import train_test_split
features = ['ChgOffDate', 'UrbanRural', 'ApprovalYear']
X=df[features]
y=df['Accept']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=101)

In [ ]:
#Using KNN
#Import KNeighborsClassifier from scikit learn.
from sklearn.neighbors import KNeighborsClassifier

In [126]:
#Create a KNN model instance with n_neighbors=
knn = KNeighborsClassifier(n_neighbors = 4)

In [127]:
#Fit this KNN model to the training data.
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=4)

In [128]:
#Predictions and Evaluations
#Let's evaluate our KNN model!
#Use the predict method to predict values using your KNN model and X_test.

pred = knn.predict(X_test)

In [129]:
#Create a confusion matrix and classification report.
from sklearn.metrics import classification_report, confusion_matrix

In [130]:
#Calculate the confusion matrix, which will help us to know the data that has 
#been correctly predicted (main diagonal = 6730), and the wrong data (secondary diagonal = 1916)
print(confusion_matrix(y_test, pred))

[[1896    6]
 [  38 7770]]


In [131]:
#Calculate model precision
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       0.98      1.00      0.99      1902
           1       1.00      1.00      1.00      7808

    accuracy                           1.00      9710
   macro avg       0.99      1.00      0.99      9710
weighted avg       1.00      1.00      1.00      9710



In [132]:
#We use this code to find a best K number to our algorithm
from sklearn.model_selection import GridSearchCV
param_grid = {'n_neighbors': np.arange(1, 10,1)} 
gs = GridSearchCV(KNeighborsClassifier(), param_grid)
gs.fit(X_train, y_train)
print("Best score: ", gs.best_score_)
print("Best params: ", gs.best_params_)
for i, max_depth in enumerate(gs.cv_results_['params']):
    print("%0.3f (+/-%0.03f) for %r" % (gs.cv_results_['mean_test_score'][i],
                                        gs.cv_results_['std_test_score'][i] * 2,
                                        max_depth))

Best score:  0.9931508334198516
Best params:  {'n_neighbors': 2}
0.986 (+/-0.016) for {'n_neighbors': 1}
0.993 (+/-0.002) for {'n_neighbors': 2}
0.993 (+/-0.002) for {'n_neighbors': 3}
0.993 (+/-0.002) for {'n_neighbors': 4}
0.993 (+/-0.002) for {'n_neighbors': 5}
0.993 (+/-0.002) for {'n_neighbors': 6}
0.993 (+/-0.002) for {'n_neighbors': 7}
0.993 (+/-0.002) for {'n_neighbors': 8}
0.993 (+/-0.002) for {'n_neighbors': 9}


In [133]:
#kNN para 9 vecinos
knn = KNeighborsClassifier(n_neighbors =4)
knn.fit(X_train, y_train)
pred = knn.predict(X_test)
print(confusion_matrix(y_test, pred))
print(classification_report(y_test, pred))

[[1896    6]
 [  38 7770]]
              precision    recall  f1-score   support

           0       0.98      1.00      0.99      1902
           1       1.00      1.00      1.00      7808

    accuracy                           1.00      9710
   macro avg       0.99      1.00      0.99      9710
weighted avg       1.00      1.00      1.00      9710



### KNN: Results

In [101]:
ids_test = df_test_nolabel['id']

features = ['ChgOffDate', 'UrbanRural', 'ApprovalYear']
X=df_test_nolabel[features]
y_pred=knn.predict(X)

In [98]:
results = df_test_nolabel[["id"]]
results["AcceptFloat"] = y_pred
results["Accept"] = results["AcceptFloat"].values.astype(int)
results.drop(['AcceptFloat'], axis=1, inplace=True)
results["id"].size

<ipython-input-98-c3a384c4651c>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  results["AcceptFloat"] = y_pred
<ipython-input-98-c3a384c4651c>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  results["Accept"] = results["AcceptFloat"].values.astype(int)
C:\Users\Alejandro\anaconda3\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_gu

7994

In [99]:
results.to_csv("resultsKNN_good.csv", index_label=False, index=False)

### Decision Tree: Training

In [ ]:
# Training and test spliting
from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.tree import DecisionTreeClassifier
from sklearn import tree

# Si se incluyen features con nan no funciona el árbol
features = ['ChgOffDate', 'UrbanRural', 'ApprovalYear']

# Transform dataframe in numpy arrays
X = df[features].values
y = df['Accept'].values

# Test set will be the 25% taken randomly
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.05, random_state=33)

# Preprocess: normalize
scaler = preprocessing.StandardScaler().fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

In [ ]:
from sklearn.model_selection import GridSearchCV
param_grid = {'max_depth': np.arange(1, 20)} 

gs = GridSearchCV(DecisionTreeClassifier(), param_grid)

gs.fit(X_train, y_train)

# summarize the results of the grid search
print("Best score: ", gs.best_score_)
print("Best params: ", gs.best_params_)

In [ ]:
# We print the score for each value of max_depth
for i, max_depth in enumerate(gs.cv_results_['params']):
    print("%0.4f (+/-%0.04f) for %r" % (gs.cv_results_['mean_test_score'][i],
                                        gs.cv_results_['std_test_score'][i] * 2,
                                        max_depth))

In [ ]:
# Evaluation
from sklearn import metrics
from sklearn.model_selection import cross_val_score, KFold, StratifiedKFold
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report
from sklearn.metrics import roc_curve
from sklearn.metrics import roc_auc_score

model = tree.DecisionTreeClassifier(criterion='gini',
                                            splitter="best",
                                            min_samples_split=3,
                                            min_samples_leaf=3,
                                            max_depth = 3,
                                            class_weight= {0:0.96,1:0.96} )

model.fit(X_train, y_train)
predicted = model.predict(X_test)
expected = y_test

# Accuracy
metrics.accuracy_score(expected, predicted)

In [ ]:
cv = KFold(n_splits=20, shuffle=False)
scores = cross_val_score(model, X, y, cv=cv)

print("Scores in every iteration", scores)
print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))

In [ ]:
y_train_pred = model.predict(X_train)
print("Accuracy in training", metrics.accuracy_score(y_train, y_train_pred))

In [ ]:
print(classification_report(y_train, y_train_pred))

In [ ]:
ids_test = df_test_nolabel['id']

features = ['ChgOffDate', 'UrbanRural', 'ApprovalYear']
X=df_test_nolabel[features]
y_pred=model.predict(X)

In [ ]:
results = df_test_nolabel[["id"]]
results["AcceptFloat"] = y_pred
results["Accept"] = results["AcceptFloat"].values.astype(int)
results.drop(['AcceptFloat'], axis=1, inplace=True)
results["id"].size

In [ ]:
results.to_csv("resultsDT_good.csv", index_label=False, index=False)

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score,f1_score


tree_clf = DecisionTreeClassifier()
tree_clf.fit(X_train,y_train)
y_pred = tree_clf.predict(X_train)

print("Training Data Set Accuracy: ", accuracy_score(y_train,y_pred))
print("Training Data F1 Score ", f1_score(y_train,y_pred))

print("Validation Mean F1 Score: ",cross_val_score(tree_clf,X_train,y_train,cv=5,scoring='f1_macro').mean())
print("Validation Mean Accuracy: ",cross_val_score(tree_clf,X_train,y_train,cv=5,scoring='accuracy').mean())

In [ ]:
df.head()

OverFitting Problem : We can see from above metrics that Training Accuracy > Test Accuracy with default settings of Decision Tree classifier. Hence, model is overfit. We will try some Hyper-parameter tuning and see if it helps.

First let's try tuning 'Max_Depth' of tree

In [ ]:
training_accuracy = []
val_accuracy = []
training_f1 = []
val_f1 = []
tree_depths = []

for depth in range(1,20):
    tree_clf = DecisionTreeClassifier(max_depth=depth)
    tree_clf.fit(X_train,y_train)
    y_training_pred = tree_clf.predict(X_train)

    training_acc = accuracy_score(y_train,y_training_pred)
    train_f1 = f1_score(y_train,y_training_pred)
    val_mean_f1 = cross_val_score(tree_clf,X_train,y_train,cv=5,scoring='f1_macro').mean()
    val_mean_accuracy = cross_val_score(tree_clf,X_train,y_train,cv=5,scoring='accuracy').mean()
    
    training_accuracy.append(training_acc)
    val_accuracy.append(val_mean_accuracy)
    training_f1.append(train_f1)
    val_f1.append(val_mean_f1)
    tree_depths.append(depth)
    

Tuning_Max_depth = {"Training Accuracy": training_accuracy, "Validation Accuracy": val_accuracy, "Training F1": training_f1, "Validation F1":val_f1, "Max_Depth": tree_depths }
Tuning_Max_depth_df = pd.DataFrame.from_dict(Tuning_Max_depth)

plot_df = Tuning_Max_depth_df.melt('Max_Depth',var_name='Metrics',value_name="Values")

fig,ax = plt.subplots(figsize=(15,5))
sns.pointplot(x="Max_Depth", y="Values",hue="Metrics", data=plot_df,ax=ax)

From above graph, we can conclude that keeping 'Max_Depth' = 3 will yield optimum Test accuracy and F1 score Optimum.

**Visulazing Decision Tree with Max Depth = 3**

In [ ]:
from sklearn import tree
tree_clf = tree.DecisionTreeClassifier(max_depth = 3)
tree_clf.fit(X_train,y_train)

In [ ]:
plt.figure(figsize=(20,10))
tree.plot_tree(tree_clf)

### GaussianNB: Training

In [196]:
#Train Test Split
#Use train_test_split to split your data into a training set and a testing set
from sklearn.model_selection import train_test_split
features= ['ChgOffDate', 'UrbanRural', 'ApprovalYear']
X=df[features]
y=df['Accept']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=101)

In [197]:
#Import Gaussian Naive Bayes model
from sklearn.naive_bayes import GaussianNB

#Create a Gaussian Classifier
model = GaussianNB()

# Train the model using the training sets
model.fit(X_train, y_train)

#Predict Output
predicted= model.predict(X_test) # 0:Overcast, 2:Mild
print ("Predicted Value:", predicted)

Predicted Value: [1 0 1 ... 1 0 1]


In [198]:
#Calculate model precision
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       0.98      1.00      0.99      1902
           1       1.00      1.00      1.00      7808

    accuracy                           1.00      9710
   macro avg       0.99      1.00      0.99      9710
weighted avg       1.00      1.00      1.00      9710



### GaussianNB: Result

In [17]:
ids_test = df_test_nolabel['id']

features = ['ChgOffDate', 'UrbanRural', 'ApprovalYear']
X=df_test_nolabel[features]
y_pred=model.predict(X)

In [ ]:
results = df_test_nolabel[["id"]]
results["AcceptFloat"] = y_pred
results["Accept"] = results["AcceptFloat"].values.astype(int)
results.drop(['AcceptFloat'], axis=1, inplace=True)
results["id"].size

In [ ]:
results.to_csv("resultsGaussianNB_good.csv", index_label=False, index=False)

### Random Forest



In [ ]:
from sklearn.model_selection import train_test_split
#On divise notre Dataset en Train set et en Test set

train_set, test_set = train_test_split(df, test_size = 0.3, random_state = 0)

In [ ]:
X = df.drop(['Accept', 'id'], axis=1)
y = df['Accept']

In [ ]:
def preprocessing(df):
    
    X = df.drop(['Accept', 'id'], axis=1)
    y = df['Accept']
    
    print(X.shape, y.shape)
    return X,y

In [ ]:
X_train, y_train = preprocessing(train_set) 

In [ ]:
X_train.columns

In [ ]:
X_test, y_test = preprocessing(test_set) 

# 4. Result of the competition
### A. Introducción
### B. Modelo Geométrico: KNN
### C. Modelo Decision Tree
### D. Modelo Gaussian NB
-----------------------------------------
### A. Introducción
Una vez disponga de un modelo, deberá usar la plataforma Kaggle para probar su modelo con datos, que combinan datos nuevos con los datos existentes. Puede realizar tantos intentos como desee. 
Razone los resultados en el cuaderno.
### B. Modelo Geométrico: KNN
Knn para 9 vecinos con todas las features preprocesadas
 
      precision    recall  f1-score   support

           0       0.28      0.28      0.28      2817
           1       0.83      0.83      0.83     11748

    accuracy                           0.72     14565
   macro avg       0.55      0.55      0.55     14565
weighted avg       0.72      0.72      0.72     14565

Best score:  0.7972748684351014
Best params:  {'n_neighbors': 9}

  Knn para 9 vecinos con 4 features preprocesadas
  
       precision    recall  f1-score   support

           0       0.36      0.08      0.14      2817
           1       0.81      0.96      0.88     11748

    accuracy                           0.79     14565
   macro avg       0.59      0.52      0.51     14565
weighted avg       0.73      0.79      0.74     14565

Best score:  0.7972748684351014
Best params:  {'n_neighbors': 9}

Knn con 4 features: 0.62095


Con 4 vecinos

                 precision    recall  f1-score   support

           0       0.98      1.00      0.99      1902
           1       1.00      1.00      1.00      7808

    accuracy                           1.00      9710
   macro avg       0.99      1.00      0.99      9710
weighted avg       1.00      1.00      1.00      9710

Con 9 vecinos

              precision    recall  f1-score   support

           0       0.98      1.00      0.99      1902
           1       1.00      1.00      1.00      7808

    accuracy                           1.00      9710
   macro avg       0.99      1.00      0.99      9710
weighted avg       1.00      1.00      1.00      9710


### C. Modelo Decision Tree

Decison tree con todas las variables (bueno, las que no tienen nan) preprocesadas 

              precision    recall  f1-score   support

           0       0.71      0.22      0.34      6892
           1       0.84      0.98      0.91     29518

    accuracy                           0.84     36410
   macro avg       0.78      0.60      0.62     36410
weighted avg       0.82      0.84      0.80     36410

Accuracy in training 0.8353474320241692

Best score:  0.8292776709695138
Best params:  {'max_depth': 6}

0.66588 en kaagle 

Decison tree con 4 variables preprocesadas

              precision    recall  f1-score   support

           0       0.71      0.20      0.31      6892
           1       0.84      0.98      0.90     29518

    accuracy                           0.83     36410
   macro avg       0.77      0.59      0.61     36410
weighted avg       0.81      0.83      0.79     36410

0.8155251893315772


Tienen mejor valor en el 0 que el knn

                precision    recall  f1-score   support

           0       0.97      1.00      0.98      8717
           1       1.00      0.99      1.00     37402

    accuracy                           0.99     46119
   macro avg       0.98      0.99      0.99     46119
weighted avg       0.99      0.99      0.99     46119

### D. Modelo Gaussian NB


              precision    recall  f1-score   support

           0       0.36      0.08      0.14      2817
           1       0.81      0.96      0.88     11748

    accuracy                           0.79     14565
   macro avg       0.59      0.52      0.51     14565
weighted avg       0.73      0.79      0.74     14565


Reusltado de kaagle 0.9932

# References

Notebooks de la asignatura